pallets/pips/src/lib.rs

```diff
@@ -809,194 +864,264 @@ decl_module! {
+        /// * `CannotSkipPip` - a given PIP has already been skipped too many times.
+        /// * `SnapshotResultTooLarge` - on len(results) > len(snapshot_queue).
+        /// * `SnapshotIdMismatch` - if:
+        ///   ```
+        ///    ∃ (i ∈ 0..SnapshotQueue.len()).
+        ///      results[i].0 ≠ SnapshotQueue[SnapshotQueue.len() - i].id
+        ///   ```
+        ///    This is protects against clearing queue while GC is voting.
+        #[weight = (1_000_000_000, DispatchClass::Operational, Pays::Yes)]
+        pub fn enact_snapshot_results(origin, results: Vec<(PipId, SnapshotResult)>) -> DispatchResult {
+            T::VotingMajorityOrigin::ensure_origin(origin)?;
 
-            // 2. New value should be valid block number.
-            let next_block = <system::Module<T>>::block_number() + 1.into();
-            let new_until = until.unwrap_or(next_block);
-            ensure!( new_until >= next_block, Error::<T>::InvalidFutureBlockNumber);
+            let max_pip_skip_count = Self::max_pip_skip_count();
+
+            <SnapshotQueue<T>>::try_mutate(|queue| {
+                let mut to_bump_skipped = Vec::new();
+                let mut to_reject = Vec::new();
+                let mut to_approve = Vec::new();
+
+                // Go over each result...
+                for (id, action) in results.iter().copied() {
+                    match queue.pop() { // ...and "zip" with the queue in reverse.
+                        // An action is missing a corresponding PIP in the queue, bail!
+                        None => return Err(Error::<T>::SnapshotResultTooLarge.into()),
+                        // The id at queue position vs. results mismatches.
+                        Some(p) if p.id != id => return Err(Error::<T>::SnapshotIdMismatch.into()),
+                        // All is right...
+                        Some(_) => {},
+                    }
+                    match action {
+                        // Make sure the PIP can be skipped and enqueue bumping of skip.
+                        SnapshotResult::Skip => {
+                            let count = PipSkipCount::get(id);
+                            ensure!(count < max_pip_skip_count, Error::<T>::CannotSkipPip);
+                            to_bump_skipped.push((id, count + 1));
+                        },
+                        // Mark PIP as rejected.
+                        SnapshotResult::Reject => to_reject.push(id),
+                        // Approve PIP.
+                        SnapshotResult::Approve => to_approve.push(id),
+                    }
+                }
 
-            // 3. Update enactment period.
-            // 3.1 Update referendum.
-            let referendum = Self::referendums(id)
-                .ok_or_else(|| Error::<T>::NoSuchProposal)?;
+                // Update skip counts.
+                for (pip_id, new_count) in to_bump_skipped.iter().copied() {
+                    PipSkipCount::insert(pip_id, new_count);
```

Comment:
We don't need to update storage every time. We can just update it once outside the loop.

Reply:
We have a map, not a single global value, so such an operation isn't defined:
```rust
        pub PipSkipCount get(fn pip_skip_count): map hasher(twox_64_concat) PipId => SkippedCount;
```